contracts/protocol/integration/amm/ArrakisUniswapV3AmmAdapter.sol

```diff
@@ -0,0 +1,277 @@
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    SPDX-License-Identifier: Apache-2.0
+*/
+
+pragma solidity 0.8.17;
+
+import "@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol";
+import "@uniswap/v3-core/contracts/interfaces/IUniswapV3Factory.sol";
+
+import "../../../interfaces/IAmmAdapter.sol";
+import "../../../interfaces/external/IArrakisVaultV1.sol";
+
+/**
+ * @title UniswapV3AmmAdapter
+ * @author Zishan Sami
+ *
+ * Adapter for Arrakis Vault representing Uniswap V3 liquidity position that encodes adding and removing liquidty
+ */
+contract ArrakisUniswapV3AmmAdapter is IAmmAdapter {
+    /* ============ State Variables ============ */
+
+    // Address of Arrakis Router contract
+    address public immutable router;
+
+    // UniswapV3 factory contract
+    IUniswapV3Factory public immutable uniV3Factory;
+
+    // Internal function string for adding liquidity
+    string internal constant ADD_LIQUIDITY =
+        "addLiquidity(address,uint256,uint256,uint256,uint256,address)";
+    // Internal function string for removing liquidity
+    string internal constant REMOVE_LIQUIDITY =
+        "removeLiquidity(address,uint256,uint256,uint256,address)";
+
+    /* ============ Constructor ============ */
+
+    /**
+     * Set state variables
+     *
+     * @param _router          Address of Arrakis Router contract
+     * @param _uniV3Factory    Address of UniswapV3 Factory contract
+     */
+    constructor(address _router, address _uniV3Factory){
+        require(_router != address(0),"_router address must not be zero address");
+        require(_uniV3Factory != address(0),"_uniV3Factory address must not be zero address");
+        router = _router;
+        uniV3Factory = IUniswapV3Factory(_uniV3Factory);
+    }
+
+    /* ============ External Getter Functions ============ */
+
+    /**
+     * Return calldata for the add liquidity call
+     *
+     * @param  _setToken                Address of the SetToken
+     * @param  _pool                    Address of liquidity token
+     * @param  _components              Token address array required to remove liquidity
+     * @param  _maxTokensIn             AmountsIn desired to add liquidity
+     * @param  _minLiquidity            Min liquidity amount to add
+     */
+    function getProvideLiquidityCalldata(
+        address _setToken,
+        address _pool,
+        address[] calldata _components,
+        uint256[] calldata _maxTokensIn,
+        uint256 _minLiquidity
+    )
+        external
+        view
+        override
+        returns (address target, uint256 value, bytes memory data)
+    {   
+        address setToken = _setToken;
+        address[] memory components = _components;
+        uint256[] memory maxTokensIn = _maxTokensIn;
+        uint256 minLiquidity = _minLiquidity;
```

Comment:
Why are we doing these reassignments ? Seems like a waste of gas no ? (For the arrays I suggest just changing the type to "memory" in the call signature if those arrays need to be edited. If they are read only then calldata is cheaper anyway.